### PR TITLE
Move `estimateSignedTxSize` to separate module

### PR DESCRIPTION
**Pull request by Anviking**
```haskell
-- |
-- Copyright: ¬© 2023 IOHK, 2023 Cardano Foundation
-- License: Apache-2.0
--
-- Module for 'signTx' and signing-related utilities for balancing.
module Cardano.Wallet.Write.Tx.Sign
    (
    -- * Signing transactions
    -- TODO: Move signTx function here

    -- * Signing-related utilities required for balancing
      estimateSignedTxSize

    , KeyWitnessCount (..)
    , estimateKeyWitnessCount
    )
    where
```

### Comments

- Depends on #4036 

### Issue Number

ADP-3081

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Sign.hs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+module Cardano.Wallet.Write.Tx.Sign
+    (
+    -- * Signing transactions
+    -- TODO: Move signTx function here
+
+    -- * Signing-related utilities required for balancing
+      estimateSignedTxSize
+
+    , KeyWitnessCount (..)
+    , estimateKeyWitnessCount
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Sign.hs:21]: Just a thought: the `estimateSignedTxSize` and `estimateKeyWitnessCount` functions seem like ideal candidates to place within the `Balance.SizeEstimation` module, or a more general `Balance.Estimation` module, given that `estimateKeyWitnessCount` is not a size estimation, per se.

Comment (Anviking) [reply lib/wallet/src/Cardano/Wallet/Write/Tx/Sign.hs:21]: I did very much want `estimateKeyWitnessCount` and `signTx` next to each other. <s>`estimateSignedTx` might make sense in `Write.Tx.SizeEstimation` though.</s> I'd suggest to revisit in another PR regardless.

Comment (Anviking) [reply lib/wallet/src/Cardano/Wallet/Write/Tx/Sign.hs:21]: > estimateSignedTx might make sense in Write.Tx.SizeEstimation though

Actually — no. `estimateSignedTx` is very different from the functions in the `SizeEstimation` module. It:
- is used by balanceTransaction, not coin selection
- doesn't actually estimate anything itself, it takes the previously estimated `KeyWitnessCount` as an argument
- is implemented using `evaluateMinimumFee`

So it should perhaps be called `evaluateSignedTxSize` instead.

Comment (Anviking) [reply lib/wallet/src/Cardano/Wallet/Write/Tx/Sign.hs:21]: > I did very much want estimateKeyWitnessCount and signTx next to each other.

The reason being that they should match. Or technically:
- `signTx` must never add more witnesses than `estimateKeyWitnessCount` estimated
- `estimateKeyWitnessCount` should never overestimate when transactions contain no native scripts

Related: https://cardanofoundation.atlassian.net/browse/ADP-2675

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Sign.hs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+            estimateWitNumForCred cred
+        _ -> 1
+      where
+        -- Does not include the key witness needed for script credentials.
+        -- They are accounted for separately in @scriptVkWitsUpperBound@.
+        estimateWitNumForCred = \case
+            Cardano.StakeCredentialByKey _ -> 1
+            Cardano.StakeCredentialByScript _ -> 0
+
+
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Sign.hs]:
```suggestion

```

**Review by jonathanknowles — APPROVED** [review]

**Pull request merged**